src/firmware/app/control/trajectory_planner_impl.c

```diff
@@ -131,29 +131,37 @@ void app_trajectory_planner_impl_modifySpeedsToMatchLongestSegmentDuration(
     for (unsigned int i = 0; i < num_elements - 1; i++)
     {
         // Check for the case that each segment duration isn't equal, so that they can be
-        // modifed to have the same duration
+        // modified to have the same duration
         if (durations1[i] > durations2[i] && displacement1[i] != 0)
         {
             const float current_speed    = speeds2[i];
-            const float displacement     = displacement2[i];
             const float desired_duration = durations1[i];
             float *final_speed_to_change = &speeds2[i + 1];
+            const float displacement     = displacement2[i];
 
             *final_speed_to_change =
                 app_trajectory_planner_impl_modifySpeedToMatchDuration(
                     current_speed, desired_duration, displacement);
+            if (*final_speed_to_change < 0)
+            {
+                *final_speed_to_change = 0;
+            }
             complete_time_profile[i + 1] = complete_time_profile[i] + desired_duration;
         }
         else if (durations2[i] > durations1[i] && displacement2[i] != 0)
         {
             const float current_speed    = speeds1[i];
-            const float displacement     = displacement1[i];
             const float desired_duration = durations2[i];
             float *final_speed_to_change = &speeds1[i + 1];
+            const float displacement     = displacement1[i];
 
             *final_speed_to_change =
                 app_trajectory_planner_impl_modifySpeedToMatchDuration(
                     current_speed, desired_duration, displacement);
+            if (*final_speed_to_change < 0)
+            {
+                *final_speed_to_change = 0;
+            }
```

Comment:
doesn't this prevent us from slowing down?

Reply:
This is absolute speed in this case, so it prevents us from going backwards. 

I think you've got the impression this is a delta-speed?

Reply:
Then the variable name is confusing because it implies change

Reply:
It is meant in the context of _changing_ the value of that speed. I'll rename to 'next_speed'